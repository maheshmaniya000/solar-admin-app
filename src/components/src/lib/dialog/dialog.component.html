<div mat-dialog-title>
  <button
    *ngIf="closeIconButtonEnabled"
    [mat-dialog-close]="undefined"
    class="close-button"
    mat-icon-button
    data-test="close-button"
  >
    <mat-icon>close</mat-icon>
  </button>
  <h3>{{ titleTranslationKey | transloco }}</h3>
</div>
<mat-divider *ngIf="dividedContent" class="top-content-divider"></mat-divider>
<mat-dialog-content [ngClass]="{ divided: dividedContent }">
  <ng-content></ng-content>
</mat-dialog-content>
<mat-divider
  *ngIf="dividedContent"
  class="bottom-content-divider"
></mat-divider>
<ng-content select="mat-dialog-actions"></ng-content>
<mat-dialog-actions *ngIf="!actionsDirective" align="end">
  <ng-container *ngFor="let action of actions | async">
    <button
      *ngIf="!action.dismissing; else dismissingButton"
      [disabled]="action.disabled"
      [mat-dialog-close]="action.payload"
      cdkFocusInitial
      color="accent"
      mat-flat-button
      [attr.data-test]="action.dataTest"
    >
      {{ action.textTranslationKey | transloco }}
    </button>
    <ng-template #dismissingButton>
      <button
        [disabled]="action.disabled"
        [mat-dialog-close]="action.payload"
        color="accent"
        mat-stroked-button
        [attr.data-test]="action.dataTest"
      >
        {{ action.textTranslationKey | transloco }}
      </button>
    </ng-template>
  </ng-container>
</mat-dialog-actions>
<ng-container *ngIf="footer">
  <mat-divider class="footer-divider"></mat-divider>
  <ng-content select="sg-dialog-footer"></ng-content>
</ng-container>
