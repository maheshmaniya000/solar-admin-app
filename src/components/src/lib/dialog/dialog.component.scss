$vertical-content-spacing: 12px;

.mat-dialog-title {
  display: flex;
  align-items: center;

  h3 {
    margin-bottom: 0;
  }
}

.mat-dialog-content {
  padding-top: $vertical-content-spacing;
  padding-bottom: $vertical-content-spacing;

  ::ng-deep .mat-divider-vertical {
    margin-top: -$vertical-content-spacing;
    margin-bottom: -$vertical-content-spacing;
  }
}

:host ::ng-deep .sg-dialog-footer {
  display: flex;
  align-items: center;
  min-height: 48px;
}
