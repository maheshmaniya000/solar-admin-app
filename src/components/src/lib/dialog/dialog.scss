@use 'variables';

$medium-dialog-vertical-spacing: 20px;
$medium-dialog-horizontal-spacing: 32px;
$medium-dialog-actions-min-height: 56px;

$large-dialog-vertical-spacing: 28px;
$large-dialog-horizontal-spacing: 48px;
$large-dialog-actions-min-height: 68px;

$mat-dialog-actions-vertical-padding: 8px;

$medium-button-height: map-get(variables.$sg-button-height, medium);
$large-button-height: map-get(variables.$sg-button-height, large);

@mixin _divider-horizontal-spacing($value) {
  margin-right: -$value;
  margin-left: -$value;
}

@mixin _dialog(
  $vertical-spacing,
  $horizontal-spacing,
  $actions-min-height,
  $button-height,
  $title-margin-bottom: null
) {
  padding: $vertical-spacing $horizontal-spacing;

  @if $title-margin-bottom {
    .mat-dialog-title {
      margin-bottom: $title-margin-bottom;
    }
  }

  .mat-dialog-content {
    margin-right: -$horizontal-spacing;
    margin-left: -$horizontal-spacing;
    padding-right: $horizontal-spacing;
    padding-left: $horizontal-spacing;
  }

  .top-content-divider,
  .bottom-content-divider {
    @include _divider-horizontal-spacing($horizontal-spacing);
  }

  .mat-dialog-actions {
    min-height: $actions-min-height;
    margin-bottom: -($vertical-spacing - $mat-dialog-actions-vertical-padding);
  }

  .sg-dialog-footer {
    margin-bottom: -$vertical-spacing;
  }

  .footer-divider {
    margin-top: $vertical-spacing;
    @include _divider-horizontal-spacing($horizontal-spacing);
  }
}

.sg-dialog-container {
  .mat-dialog-container {
    .close-button {
      margin-right: 6px;
      margin-left: -6px;
    }

    @include _dialog(
      $medium-dialog-vertical-spacing,
      $medium-dialog-horizontal-spacing,
      $medium-dialog-actions-min-height,
      $medium-button-height
    );
  }

  &.large {
    .mat-dialog-container {
      @include _dialog(
        $large-dialog-vertical-spacing,
        $large-dialog-horizontal-spacing,
        $large-dialog-actions-min-height,
        $large-button-height,
        $large-dialog-vertical-spacing
      );
    }

    .mat-dialog-actions button {
      height: $large-button-height;
    }

    &.featured {
      .mat-dialog-title,
      .mat-dialog-content,
      .mat-dialog-actions,
      .sg-dialog-footer {
        justify-content: center;
      }

      .mat-dialog-title {
        position: relative;
        margin-bottom: 0;
      }

      .close-button {
        position: absolute;
        top: -$large-dialog-vertical-spacing + 6px;
        left: -$large-dialog-horizontal-spacing + 12px;
      }

      .mat-dialog-content {
        text-align: center;
      }
    }
  }
}
