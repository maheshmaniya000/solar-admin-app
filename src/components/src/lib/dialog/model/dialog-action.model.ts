import { Observable } from 'rxjs';

export interface DialogAction<T = any> {
  dismissing: boolean;
  textTranslationKey: string;
  payload: T;
  disabled?: boolean;
  dataTest: string;
}

// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays
type TypesToDialogActions<Types extends any[]> = { [K in keyof Types]: DialogAction<Types[K]> };

export type DialogActions<PayloadTypes extends any[] = any[]> =
  | TypesToDialogActions<PayloadTypes>
  | Observable<TypesToDialogActions<PayloadTypes>>;
