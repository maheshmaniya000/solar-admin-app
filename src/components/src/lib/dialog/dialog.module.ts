import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import {
  MatDialogActions,
  MatDialogClose,
  MatDialogModule,
  MAT_DIALOG_DEFAULT_OPTIONS
} from '@angular/material/dialog';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule } from '@angular/material/icon';
import { TranslocoModule } from '@ngneat/transloco';

import { DialogComponent } from './dialog.component';
import { DialogConstants } from './dialog.constants';
import { DialogFooterDirective } from './footer-content/dialog-footer.directive';
import { StandardDialogComponent } from './standard-dialog/standard-dialog.component';
import { SuccessDialogContentComponent } from './success-dialog/success-dialog-content.component';

@NgModule({
  declarations: [
    StandardDialogComponent,
    DialogComponent,
    DialogFooterDirective,
    SuccessDialogContentComponent
  ],
  imports: [
    CommonModule,
    MatButtonModule,
    MatDialogModule,
    MatDividerModule,
    MatIconModule,
    TranslocoModule
  ],
  exports: [
    MatDialogActions,
    StandardDialogComponent,
    DialogComponent,
    DialogFooterDirective,
    MatDialogClose
  ],
  providers: [
    {
      provide: MAT_DIALOG_DEFAULT_OPTIONS,
      useValue: DialogConstants.defaultConfigs.medium
    }
  ]
})
export class DialogModule {}
