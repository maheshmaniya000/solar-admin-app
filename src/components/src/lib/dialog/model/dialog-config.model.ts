import { MatDialogConfig } from '@angular/material/dialog';

export type DialogConfig<D = any> = Pick<MatDialogConfig<D>, 'data' | 'disableClose'>;
