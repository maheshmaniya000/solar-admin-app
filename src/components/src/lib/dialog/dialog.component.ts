import {
  AfterContentChecked,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChild,
  Input,
  OnChanges,
  OnInit
} from '@angular/core';
import { MatDialogActions } from '@angular/material/dialog';
import { isNil } from 'lodash-es';
import { isObservable,  of } from 'rxjs';

import { DialogConstants } from './dialog.constants';
import { DialogFooterDirective } from './footer-content/dialog-footer.directive';
import { DialogActions } from './model/dialog-action.model';

@Component({
  selector: 'sg-dialog',
  templateUrl: './dialog.component.html',
  styleUrls: ['./dialog.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class DialogComponent implements OnInit, OnChanges, AfterContentChecked {
  @Input() titleTranslationKey: string;
  @Input() dividedContent = false;
  @Input() closeIconButtonEnabled = false;
  @Input() actions: DialogActions;
  @ContentChild(MatDialogActions) actionsDirective: MatDialogActions;
  @ContentChild(DialogFooterDirective)
  footer: DialogFooterDirective;

  constructor(private readonly changeDetectorRef: ChangeDetectorRef) {}

  ngOnChanges(): void {
    if (isNil(this.actions)) {
      this.actions = of([
        DialogConstants.actions.cancel,
        DialogConstants.actions.ok
      ]);
    }
    this.actions = isObservable(this.actions) ? this.actions : of(this.actions);
  }

  ngOnInit(): void {
    this.ngOnChanges();
  }

  ngAfterContentChecked(): void {
    this.changeDetectorRef.markForCheck();
  }
}
