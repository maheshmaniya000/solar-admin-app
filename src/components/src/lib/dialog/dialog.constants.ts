import { MatDialogConfig } from '@angular/material/dialog';

import { DialogAction } from './model/dialog-action.model';
import { DialogKind } from './model/dialog-kind.model';

type DialogActionKind = 'cancel' | 'ok' | 'delete' | 'change';

export class DialogConstants {
  static readonly confirmAction: Omit<DialogAction<true>, 'textTranslationKey' | 'dataTest'> = {
    dismissing: false,
    payload: true
  };

  static readonly actions: Record<DialogActionKind, DialogAction<boolean>> = {
    cancel: {
      dismissing: true,
      textTranslationKey: 'common.action.cancel',
      payload: false,
      dataTest: 'cancel-action-button'
    },
    ok: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.ok',
      dataTest: 'ok-action-button'
    },
    delete: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.delete',
      dataTest: 'delete-action-button'
    },
    change: {
      ...DialogConstants.confirmAction,
      textTranslationKey: 'common.action.change',
      dataTest: 'change-action-button'
    }
  };

  private static readonly sgDialogContainerClass = 'sg-dialog-container';
  private static readonly sgLargeDialogContainerClass = 'large';
  private static readonly sgFeaturedDialogContainerClass = 'featured';
  private static readonly defaultMediumDialogWidth = '608px';
  private static readonly defaultLargeDialogWidth = '640px';

  static readonly defaultConfigs: Record<DialogKind, MatDialogConfig> = {
    medium: {
      ...new MatDialogConfig(),
      panelClass: DialogConstants.sgDialogContainerClass,
      width: DialogConstants.defaultMediumDialogWidth
    },
    large: {
      ...new MatDialogConfig(),
      panelClass: [
        DialogConstants.sgDialogContainerClass,
        DialogConstants.sgLargeDialogContainerClass
      ],
      width: DialogConstants.defaultLargeDialogWidth
    },
    featured: {
      ...new MatDialogConfig(),
      panelClass: [
        DialogConstants.sgDialogContainerClass,
        DialogConstants.sgLargeDialogContainerClass,
        DialogConstants.sgFeaturedDialogContainerClass
      ],
      width: DialogConstants.defaultLargeDialogWidth
    }
  };

  static createStandardDialogDefaultActions<T>(
    okActionPayload?: T,
    okActionDisabled: boolean = false
  ): [DialogAction<undefined>, DialogAction<T>] {
    return [
      { ...DialogConstants.actions.cancel, payload: undefined },
      {
        ...DialogConstants.actions.ok,
        payload: okActionPayload,
        disabled: okActionDisabled
      }
    ];
  }
}
