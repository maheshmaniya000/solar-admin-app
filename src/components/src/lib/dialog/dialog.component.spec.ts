import { ElementRef } from '@angular/core';
import {
  MatDialog,
  MatDialogActions,
  MatDialogClose,
  MatDialogContent,
  MatDialogTitle
} from '@angular/material/dialog';
import {
  createHostFactory,
  SpectatorHost
} from '@ngneat/spectator/jest';
import { TranslocoTestingModule } from '@ngneat/transloco';
import { MockDirective } from 'ng-mocks';
import {
  Observable,
  of
} from 'rxjs';

import { toDataTestAttributeSelector } from '../utils/test/to-data-test-attribute-selector';
import { DialogComponent } from './dialog.component';
import { DialogFooterDirective } from './footer-content/dialog-footer.directive';
import { DialogAction } from './model/dialog-action.model';

describe('DialogComponent', () => {
  describe('component', () => {
    const createHost = createHostFactory({
      component: DialogComponent,
      declarations: [
        DialogFooterDirective,
        MatDialogTitle,
        MatDialogContent,
        MatDialogActions,
        MockDirective(MatDialogClose)
      ],
      shallow: true,
      mocks: [MatDialog],
      imports: [TranslocoTestingModule]
    });
    let spectator: SpectatorHost<DialogComponent>;

    it('should contain given title text in the title', () => {
      spectator = createHost(`<sg-dialog titleTranslationKey="some.title"></sg-dialog>`);
      expect(
        spectator.query(MatDialogTitle, { read: ElementRef }).nativeElement
      ).toHaveDescendantWithText({ selector: 'h3', text: 'en.some.title' });
    });

    it('should contain transcluded content', () => {
      spectator = createHost(`<sg-dialog>Transcluded content</sg-dialog>`);
      expect(
        spectator.query(MatDialogContent, { read: ElementRef }).nativeElement
      ).toHaveText('Transcluded content');
    });

    it('should contain close dialog button with undefined payload in the title', () => {
      spectator = createHost(
        `<sg-dialog [closeIconButtonEnabled]="true"></sg-dialog>`
      );
      const closeButton = spectator.query(
        toDataTestAttributeSelector('close-button')
      );
      const matDialogCloseDirective = spectator.query(MatDialogClose);
      const matDialogCloseDirectiveElementRef = spectator.query(
        MatDialogClose,
        { read: ElementRef }
      );

      expect(
        spectator.query(MatDialogTitle, { read: ElementRef }).nativeElement
      ).toHaveDescendant(closeButton);
      expect(closeButton).toBe(matDialogCloseDirectiveElementRef.nativeElement);
      expect(matDialogCloseDirective.dialogResult).toBe(undefined);
    });

    it('should contain divided content', () => {
      spectator = createHost(`<sg-dialog [dividedContent]="true"></sg-dialog>`);
      expect(spectator.query('.top-content-divider')).toExist();
      expect(spectator.query('.bottom-content-divider')).toExist();
      expect(
        spectator.query(MatDialogContent, { read: ElementRef }).nativeElement
      ).toHaveClass('divided');
    });

    describe('dialog actions', () => {
      it(`should match snapshot of a dialog with default actions template
 created from non-observable default actions`, () => {
        spectator = createHost(`<sg-dialog></sg-dialog>`);
        expect(spectator.fixture).toMatchSnapshot();
      });

      it('should contain transcluded custom actions template', () => {
        spectator = createHost(`
      <sg-dialog>
        <mat-dialog-actions>Custom actions template</mat-dialog-actions>
      </sg-dialog>`);
        expect(
          spectator.query(MatDialogActions, { read: ElementRef }).nativeElement
        ).toHaveText('Custom actions template');
      });

      describe('custom input actions', () => {
        const actions = [
          {
            textTranslationKey: 'dismiss',
            dismissing: true,
            payload: false,
            disabled: true,
            dataTest: 'dismiss-action-button'
          },
          {
            textTranslationKey: 'ok.subscribe',
            dismissing: false,
            payload: { subscriptionEmail: 'name@solargis.com' },
            dataTest: 'ok-subscribe-action-button'
          },
          {
            textTranslationKey: 'ok.butDoNotsubscribe',
            dismissing: false,
            payload: true,
            dataTest: 'ok-but-do-not-subscribe-action-button'
          }
        ];
        const actions$: Observable<DialogAction[]> = of(actions);

        beforeEach(() => {
          spectator = createHost(
            `<sg-dialog [actions]="actions"></sg-dialog>`,
            {
              hostProps: {
                actions: actions$
              }
            }
          );
        });

        it('should contain buttons with correctly bound dialog payload attributes', () => {
          const matDialogCloseDirectives = spectator.queryAll(MatDialogClose);
          const matDialogCloseDirectiveElementRefs = spectator.queryAll(
            MatDialogClose,
            { read: ElementRef }
          );
          actions.forEach(({ dataTest, payload }, index) => {
            expect(spectator.query(toDataTestAttributeSelector(dataTest))).toBe(
              matDialogCloseDirectiveElementRefs[index].nativeElement
            );
            expect(matDialogCloseDirectives[index].dialogResult).toBe(payload);
          });
        });

        it(`should match snapshot of a dialog with actions template created
 from observable actions input`, () => {
          expect(spectator.fixture).toMatchSnapshot();
        });
      });
    });

    it('should add and remove transcluded footer with divider', () => {
      spectator = createHost(
        `
      <sg-dialog>
        <sg-dialog-footer *ngIf="footerDisplayed">Transcluded footer content</sg-dialog-footer>
      </sg-dialog>
    `,
        {
          hostProps: {
            footerDisplayed: true
          }
        }
      );
      expect(spectator.query('.footer-divider')).toExist();
      expect(spectator.query('.sg-dialog-footer')).toHaveText(
        'Transcluded footer content'
      );
      spectator.setHostInput({ footerDisplayed: false });
      expect(spectator.query('.footer-divider')).not.toExist();
      expect(spectator.query('.sg-dialog-footer')).not.toExist();
    });
  });
});
