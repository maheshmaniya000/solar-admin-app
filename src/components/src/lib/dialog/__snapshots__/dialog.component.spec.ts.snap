// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogComponent component dialog actions custom input actions should match snapshot of a dialog with actions template created
 from observable actions input 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  actions={[Function Observable]}
>
  <sg-dialog>
    <div
      class="mat-dialog-title"
      mat-dialog-title=""
    >
      <h3>
      </h3>
    </div>
    <mat-dialog-content
      class="mat-dialog-content"
    />
    <mat-dialog-actions
      align="end"
      class="mat-dialog-actions"
    >
      <button
        color="accent"
        data-test="dismiss-action-button"
        disabled=""
        mat-stroked-button=""
      >
         en.dismiss 
      </button>
      <button
        cdkfocusinitial=""
        color="accent"
        data-test="ok-subscribe-action-button"
        mat-flat-button=""
      >
         en.ok.subscribe 
      </button>
      <button
        cdkfocusinitial=""
        color="accent"
        data-test="ok-but-do-not-subscribe-action-button"
        mat-flat-button=""
      >
         en.ok.butDoNotsubscribe 
      </button>
    </mat-dialog-actions>
  </sg-dialog>
</ng-component>
`;

exports[`DialogComponent component dialog actions should match snapshot of a dialog with default actions template
 created from non-observable default actions 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <sg-dialog>
    <div
      class="mat-dialog-title"
      mat-dialog-title=""
    >
      <h3>
      </h3>
    </div>
    <mat-dialog-content
      class="mat-dialog-content"
    />
    <mat-dialog-actions
      align="end"
      class="mat-dialog-actions"
    >
      <button
        color="accent"
        data-test="cancel-action-button"
        mat-stroked-button=""
      >
         en.common.action.cancel 
      </button>
      <button
        cdkfocusinitial=""
        color="accent"
        data-test="ok-action-button"
        mat-flat-button=""
      >
         en.common.action.ok 
      </button>
    </mat-dialog-actions>
  </sg-dialog>
</ng-component>
`;
